src/box/replication.cc

```diff
@@ -500,6 +558,8 @@ replica_on_applier_state_f(struct trigger *trigger, void *event)
 		 * has been cancelled. Assume synced.
 		 */
 		replica_on_applier_sync(replica);
+		if (replica->is_applier_healthy)
+			replica_on_applier_off(replica);
 		break;
 	case APPLIER_STOPPED:
```

Comment:
> Why don't you check "off" for APPLIER_STOPPED state?

Because it's checked inside `replica_on_applier_disconnect`.

If I understood you correctly, you wanted something similar to this:
```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index b8a5cf852..85226ce61 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -395,6 +395,22 @@ replica_clear_applier(struct replica *replica)
                replica_on_applier_off(replica);
 }
 
+static void
+replica_check_applier_health(struct replica *replica)
+{
+       struct applier *applier = replica->applier;
+       assert(applier != NULL);
+       if (applier->state == APPLIER_FOLLOW) {
+               replica_on_applier_follow(replica);
+       } else if ((applier->state == APPLIER_DISCONNECTED ||
+                  applier->state == APPLIER_STOPPED ||
+                  applier->state == APPLIER_LOADING ||
+                  applier->state == APPLIER_OFF) &&
+                  replica->is_applier_healthy) {
+               replica_on_applier_off(replica);
+       }
+}
+
 static void
 replica_on_applier_sync(struct replica *replica)
 {
@@ -504,7 +520,6 @@ replica_on_applier_disconnect(struct replica *replica)
        case APPLIER_SYNC:
                assert(replicaset.applier.synced > 0);
                replicaset.applier.synced--;
-               replica_on_applier_off(replica);
                FALLTHROUGH;
        case APPLIER_CONNECTED:
                assert(replicaset.applier.connected > 0);
@@ -530,6 +545,7 @@ replica_on_applier_state_f(struct trigger *trigger, void *event)
        (void)event;
        struct replica *replica = container_of(trigger,
                        struct replica, on_applier_state);
+       replica_check_applier_health(replica);
        switch (replica->applier->state) {
        case APPLIER_INITIAL_JOIN:
                replicaset.is_joining = true;
@@ -553,7 +569,6 @@ replica_on_applier_state_f(struct trigger *trigger, void *event)
                break;
        case APPLIER_FOLLOW:
                replica_on_applier_sync(replica);
-               replica_on_applier_follow(replica);
                break;
        case APPLIER_OFF:
                /*
@@ -562,8 +577,6 @@ replica_on_applier_state_f(struct trigger *trigger, void *event)
                 * has been cancelled. Assume synced.
                 */
                replica_on_applier_sync(replica);
-               if (replica->is_applier_healthy)
-                       replica_on_applier_off(replica);
                break;
        case APPLIER_STOPPED:
                /* Unrecoverable error. */
```

Applied.